lib/liquid_voting/delegations.ex

```diff
@@ -185,19 +186,49 @@ defmodule LiquidVoting.Delegations do
   def upsert_delegation(%{delegator_id: delegator_id, delegate_id: delegate_id} = attrs) do
     proposal_url = Map.get(attrs, :proposal_url)
 
-    Delegation
-    |> where(delegator_id: ^delegator_id)
-    |> Repo.all()
-    |> resolve_conflicts(delegate_id, proposal_url)
+    with {:ok} <- check_vote_conflict(delegator_id, proposal_url) do
+      Delegation
+      |> where(delegator_id: ^delegator_id)
+      |> Repo.all()
+      |> resolve_conflicts(delegate_id, proposal_url)
+      |> case do
+        {:ok, delegations} ->
+          delegations
+          |> find_similar_delegation_or_return_new_struct(proposal_url)
+          |> Delegation.changeset(attrs)
+          |> Repo.insert_or_update()
+
+        {:error, %{message: message, details: details}} ->
+          {:error, %{message: message, details: details}}
+      end
+    end
+  end
+
+  # Checks for a conflicting vote, in the case where a proposal-specific delegation creation is attempted.
+  #
+  # Returns {:ok} if delegation creation is for a global delegation.
+  #
+  # Or returns {:ok} if delegation creation is for a proposal delegation & no conflicting vote is found.
+  # Returns an error, if a conflicting vote is found.
+  defp check_vote_conflict(_delegator_id, _proposal_url = nil) do
+    {:ok}
+  end
+
+  defp check_vote_conflict(delegator_id, proposal_url) do
+    Vote
+    |> where(participant_id: ^delegator_id, proposal_url: ^proposal_url)
+    |> Repo.one()
```

Comment:
For better separation of concerns, I'd move this query to a function under `Voting`, say `Voting.get_vote_by_participant_id`